magma/primitives/memory.py

```diff
@@ -0,0 +1,105 @@
+from hwtypes import BitVector
+
+from magma.bit import Bit
+from magma.bits import Bits
+from magma.bitutils import clog2
+from magma.clock import Clock, Enable
+from magma.clock_io import ClockIO
+from magma.conversions import from_bits, as_bits
+from magma.generator import Generator2
+from magma.interface import IO
+from magma.primitives.register import Register
+from magma.t import In, Out, Kind
+
+
+class CoreIRMemory(Generator2):
+    def __init__(self, depth, width, init=None):
+        self.name = f"coreir_mem{depth}x{width}"
+        addr_width = clog2(depth)
+        self.io = IO(
+            raddr=In(Bits[addr_width]),
+            rdata=Out(Bits[width]),
+            waddr=In(Bits[addr_width]),
+            wdata=In(Bits[width]),
+            clk=In(Clock),
+            wen=In(Enable)
+        )
+        self.verilog_name = "coreir_mem"
+        self.coreir_name = "mem"
+        self.coreir_lib = "coreir"
+        self.coreir_genargs = {"width": width, "depth": depth,
+                               "has_init": init is not None}
+        self.coreir_configargs = {}
+        if init is not None:
+            raise NotImplementedError("CoreIR's memory primitive does not "
+                                      "compile init arg to verilog yet")
+            # self.coreir_configargs["init"] = <convert to json arg>
+
+        def _simulate(self, value_store, state_store):
+            cur_clk = value_store.get_value(self.clk)
+
+            if not state_store:
+                state_store['mem'] = [
+                    BitVector[width](0) for _ in range(depth)
+                ]
+                state_store['prev_clk'] = cur_clk
+
+            prev_clk = state_store['prev_clk']
+            clk_edge = cur_clk and not prev_clk
+            rdata = value_store.get_value(self.rdata)
+
+            if clk_edge:
+                index = int(
+                    BitVector[addr_width](value_store.get_value(self.raddr))
+                )
+                rdata = state_store['mem'][index].as_bool_list()
+
+            if clk_edge:
+                if value_store.get_value(self.wen):
+                    index = int(
+                        BitVector[addr_width](value_store.get_value(self.waddr))
+                    )
+                    state_store['mem'][index] = BitVector[width](
+                        value_store.get_value(self.wdata)
+                    )
+
+            state_store['prev_clk'] = cur_clk
+            value_store.set_value(self.rdata, rdata)
+
+        self.simulate = _simulate
```

Comment:
I think we can just directly make `_simulate` a member function on the class? (or a classmethod maybe?).

Reply:
I tried that before but it didn't work, perhaps class method will work, didn't look deeply into why we can't just `def ` it like a normal member